server/src/main/java/org/elasticsearch/action/admin/cluster/bootstrap/TransportGetDiscoveredNodesAction.java

```diff
@@ -124,4 +125,22 @@ public String toString() {
             });
         }
     }
+
+    private static boolean checkWaitRequirements(GetDiscoveredNodesRequest request, Set<DiscoveryNode> nodes) {
+        Set<String> requiredNodes = new HashSet<>(request.getRequiredNodes());
+        for (final DiscoveryNode node : nodes) {
+            boolean matchedAddress = requiredNodes.remove(node.getAddress().toString());
+            boolean matchedName = requiredNodes.remove(node.getName());
+            if (matchedAddress && matchedName) {
```

Comment:
Ok, could we also match on the IP address part of the transport address (`TransportAddress#getAddress()`) (with tests, obvs)?

This opens up a slight complexity in this check-for-duplicates, because two nodes may share an IP address, so we need to consider this (i.e. have a test for it):

```
{node1}{10.20.30.40:9301}
{node2}{10.20.30.40:9302}
```

We should reject `requiredNodes` being `"node1", "10.20.30.40"` because `10.20.30.40` matches both, but the obvious change here wouldn't always pick it up.